### PR TITLE
Optimize implementations of FromIterator and Extend for Vec

### PR DESCRIPTION
**Pull request by mzabaluev**
Instead of a fast branch with a sized iterator falling back to a potentially poorly optimized iterate-and-push loop, a single efficient loop can serve all cases.

In my benchmark runs, I see some good gains, but also some regressions, possibly due to different inlining choices by the compiler. YMMV.

### CODE REVIEW

**Comment by Gankro** [comment]
Can you post your experimental results.

**Comment by mahkoh** [comment]
Does it optimize to a memcpy if the src is contiguous? If not then it will have to be changed anyway.

File: src/libcollections/vec.rs [hunk 1]
```diff
@@ -1484,11 +1468,34 @@ impl<'a, T> IntoIterator for &'a mut Vec<T> {
...
+
+impl<T> Vec<T> {
+    fn extend_desugared<I: Iterator<Item=T>>(&mut self, mut iterator: I) {
+        // This function should be the moral equivalent of:
+        //
+        //      for item in iterator {
+        //          self.push(item);
+        //      }
+        loop {
+            match iterator.next() {
```
Comment (pczarn) [inline src/libcollections/vec.rs]: Can you use `while let` here?

Comment (mzabaluev) [inline src/libcollections/vec.rs]: I can, good point.

**Comment by mzabaluev** [comment]
Results of benchmarking on i686: https://gist.github.com/mzabaluev/df41c2f50464416b0a26

tl;dr: Notable losers are (observed consistently over multiple bench runs):

* bit::bit_vec_bench::bench_bit_vec_big_iter
* slice::bench::sort_big_random_small
* slice::bench::zero_1kb_from_elem
* vec::tests::bench_extend_0000_1000
* vec::tests::bench_extend_1000_1000
* vec::tests::bench_from_elem_0100
* vec::tests::bench_from_iter_0010
* vec::tests::bench_from_iter_0100

Winners:

* string::tests::bench_push_char_two_bytes
* string::tests::from_utf8_lossy_100_invalid
* vec::tests::bench_extend_0000_0010
* vec::tests::bench_extend_0000_0100
* vec::tests::bench_from_elem_1000
* vec::tests::bench_from_iter_1000

Overall, I don't think too much trust should be put in microbenchmarks repeatedly testing one call-site optimized function.

**Comment by huonw** [comment]
How much is "winning" and "losing'"?

**Comment by mzabaluev** [comment]
@mahkoh These changes should not change the performance with exact-length iterators, in which case it's as close to `reserve` + type-aligned memcpy as the optimizer can make it. The main purpose is to provide comparable efficiency for size-unaware or pessimistic iterators.

**Comment by mzabaluev** [comment]
@huonw I've added percentages to that comment.

**Comment by pczarn** [comment]
Two of these methods have `#[inline]`, but `extend_desugared` does not; are benchmarks affected? Is there some benefit to having `extend_desugared` vs calling `vector.extend(iterator)`? Does `extend_desugared` change inlining choices, too?

**Comment by mzabaluev** [comment]
I actually ran benchmarks once with `#[inline]` on `extend_desugared`, but the results did not look more different than some jitter. I don't think `#[inline]` has any effect on generic methods, as they are always available in metadata, so the compiler can always choose to inline them across crates. Even less so for unit tests.

**Comment by mzabaluev** [comment]
There should be more benchmarks with extending from sizeless/pessimistic iterators. Any good candidates in libcollections?

**Comment by Gankro** [comment]
BTreeMap's `range` iter, VecMap's iter, and bitvset's iter have no idea how many elements they contain, but they're all also doing a non-trivial amount of work to actually yield those elements.

**Comment by mzabaluev** [comment]
@pczarn Without looking too closely into the results (I get easily frustrated with the long build times of libstd and stdtest), I assume the optimizer has more "reason" to share `extend_desugared` if it is used in various places, unless the tradeoff between cache/branch predictor efficiency and removing the overhead of calling leans towards the latter.

**Comment by steveklabnik** [comment]
How do we all feel about this PR today?

**Comment by Gankro** [comment]
Oh geez, this slipped *right* through the cracks! I don't have a great gut on this since it seems to just be shuffling perf around. r? @huonw

File: src/libcollections/vec.rs [hunk 2]
```diff
@@ -1410,42 +1410,26 @@ impl<T> ops::DerefMut for Vec<T> {
...
-            unsafe {
-                ptr::write(vector.get_unchecked_mut(len), element);
-                vector.set_len(len + 1);
-            }
-        }
-
-        if vector.len() == vector.capacity() {
-            for element in iterator {
-                vector.push(element);
+        let mut vector = match iterator.next() {
```
Comment (huonw) [inline src/libcollections/vec.rs]: Hm, does this actually improve performance, over the simpler:

```rust
let mut vector = Vec::with_capacity(iterator.size_hint().0);
vector.extend_desugared(iterator);
vector
```

Comment (mzabaluev) [inline src/libcollections/vec.rs]: This avoids [a branch](https://github.com/rust-lang/rust/pull/22681/files#diff-77adadec35cb5b03d4933f83754de940R1484) that is present in `extend_desugared`. That branch tends to be not taken, but in the first iteration the vector is always expanded.

Comment (huonw) [inline src/libcollections/vec.rs]: Hm, why it does it avoid that branch? It seems that if the iterator has a non-zero size hint the `with_capacity` will ensure that it isn't taken, and in the case that the iterator has a zero-size hint it seems both with and without this are equally bad off?

Basically, I'm asking if this was noticeable in practice.

Comment (mzabaluev) [inline src/libcollections/vec.rs]: You are right, and it was somehow lost on me that `Vec::with_capacity(1)` allocates exactly one element. I should simplify the code to your suggestion; the performance difference will likely be negligible.

Comment (mzabaluev) [inline src/libcollections/vec.rs]: With the suggested change, `vec::bench_from_iter_0000` regresses
from 29 ns/iter (+/- 2) to 87 ns/iter (+/- 48) in my testing
(rebased against commit 3dbfa74305e6cc347ffdd902a680bb3b9ee2470a), the other benchmarks seemingly unaffected. I wonder if it can be considered an edge case.

**Comment by Gankro** [comment]
@huonw @mzabaluev What's up with this?

**Comment by Gankro** [comment]
Closing due to inactivity.

**Comment by mzabaluev** [comment]
@Gankro Ah sorry, I let your request slip by. I believe the branch is good as it is. @huonw's suggestion, while good from code clarity point of view, resulted in some performance degradation as mentioned in the discussion above.

**Comment by Gankro** [comment]
@huonw Any reason not to merge this?

**Comment by Gankro** [comment]
huon's busy

File: src/libcollections/vec.rs [hunk 3]
```diff
@@ -1484,11 +1468,27 @@ impl<'a, T> IntoIterator for &'a mut Vec<T> {
...
+    fn extend_desugared<I: Iterator<Item=T>>(&mut self, mut iterator: I) {
+        // This function should be the moral equivalent of:
+        //
+        //      for item in iterator {
+        //          self.push(item);
+        //      }
+        while let Some(element) = iterator.next() {
+            let len = self.len();
+            if len == self.capacity() {
+                let (lower, _) = iterator.size_hint();
```
Comment (Gankro) [inline src/libcollections/vec.rs]: Calling size_hint in a loop seems *really* bad. This is not necessarily a straight-forward or cheap method. Is hoisting it out not worth it in your testing?

Comment (huonw) [inline src/libcollections/vec.rs]: It is called exponentially rarely, so it seems fine?

Comment (Gankro) [inline src/libcollections/vec.rs]: Oh derp, sure.

**Comment by Gankro** [comment]
At worst I like this better than the old code. r+
